equinox/nn/_attention.py

```diff
@@ -325,6 +401,42 @@ def __call__(
                 raise ValueError(
                     "process_heads must not change the shape of the heads."
                 )
+        state_length = None
+        if state is None:
+            causal_mask_offset = 0
+        else:
+            if self.kv_cache is None:
+                raise ValueError(
+                    "State was provided, but cannot use autoregressive decoding "
+                    "without specifying "
+                    "`MultiheadAttention(..., kv_cache=...)`. "
+                    "See `equinox.nn.StandardKVCache` for an example."
+                )
+
+            key_state, value_state, index, state = self.kv_cache(
+                state, key_heads, value_heads
+            )
+            _check_kv_shapes(key_state, value_state, key_heads, value_heads)
+            state_length = key_state.shape[0]
+
+            causal_mask_offset = index - kv_seq_length
```

Comment:
I think we can skip this subtraction by removing the corresponding addition inside of `StandardKVCache`:

```python
        state = state.set(self.autoregressive_index, (key_state, value_state, index + kv_seq_length))
        return key_state, value_state, index, state
```
(Note that I've still done the addition when updating `state` though!)